### PR TITLE
Initial Dart Web run configuration commit, currently disabled in the plugin.xml

### PR DESCRIPTION
**Pull request by jwren**
- the new run configuration is disabled in the `plugin.xml`,
- debugging does not yet work

@alexander-doroshko I am happy to wait until everything is farther along, or to create some smaller PRs along the way for history and review.

### CODE REVIEW

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  public DartWebdevConfigurationType() {
+    super("DartRemoteDebugConfigurationType",
+          DartBundle.message("webdev.debug.configuration.name"),
+          DartBundle.message("webdev.debug.configuration.description"),
+          NotNullLazyValue.createValue(() -> DartIcons.Dart_16));
+    addFactory(new ConfigurationFactory(this) {
+      @NotNull
+      @Override
+      public String getName() {
+        return "Dart Web"; // compatibility
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java]: That's the first version of the new run configuration, there's nothing we need to be compatible with :)

File: Dart/src/com/jetbrains/lang/dart/DartBundle.properties [hunk 2]
```diff
@@ -98,10 +98,17 @@ fields.to.generate.setters=Select Fields to Generate Setters
...
 dart.implement.method=Choose method to implement
+
 runner.command.line.configuration.name=Dart Command Line App
 runner.command.line.configuration.description=Dart Command Line Application
+
 remote.debug.configuration.name=Dart Remote Debug
 remote.debug.configuration.description=Debug Remote Dart Command Line Application
+
+webdev.debug.configuration.name=Dart Web
+webdev.debug.configuration.description=Debug Dart Webdev Application
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/DartBundle.properties]: AFAIU, webdev is just a tool that helps to deal with Dart web applications. Does 'Webdev Application' make any sense?

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+      @NotNull
+      @Override
+      public RunConfiguration createTemplateConfiguration(@NotNull final Project project) {
+        return new DartWebdevConfiguration(project, DartWebdevConfigurationType.this, "Dart Web");
+      }
+
+      @Override
+      public boolean isApplicable(@NotNull final Project project) {
+        return FileTypeIndex.containsFileOfType(HtmlFileType.INSTANCE, GlobalSearchScope.projectScope(project));
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java]: DartFileType

**Comment by alexander-doroshko** [comment]
Thanks for the start! I think, we'll merge it later, when it's closer to a working state. I guess the main chunk of work is ahead. Daemon management and all the stuff. Also, I expect a conflict with JavaScriptDebugConfigurationProducer that we'll need to resolve somehow (maybe `RunConfigurationProducer.shouldReplace` will help).

File: Dart/src/com/jetbrains/lang/dart/DartBundle.properties (hunk 2 above)
Comment (jwren) [reply Dart/src/com/jetbrains/lang/dart/DartBundle.properties]: Good point, updated to `Debug Dart Webdev`, I was just trying to be consistent... `Debug using Dart Webdev` would also be fine.

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java (hunk 1 above)
Comment (jwren) [reply Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java]: Removed

**Comment by jwren** [comment]
- Rebased on top of the latest master branch
- Updated with your comments

I'll try to keep this branch merge-able, and will keep the future commits not squashed into this initial commit so you can track the changes.

**Comment by jwren** [comment]
@alexander-doroshko 

Some notes on the current status:

- The run configuration is still disabled in the `plugin.xml`,
- This run config worked locally today (Oct 4th) with webdev version `2.5.2-dev`, to use this version I modified the command line generated to run `webdev.dart` locally,
- Expression evaluation doesn't yet work, we should disable the action when doing a launch of this run config type,
- I modified the trying to connect message to not include the connect url as we don't know ahead of time what the connect url will be,
- The Windows issues, https://github.com/dart-lang/webdev/issues/709, will be fixed soon
- The Flutter plugin has removed the timeout when attempting to connect the debugger with the idea that users can cancel or wait as appropriate, we may want to consider the same change in the Dart plugin, or perhaps just for this launch configuration.
- The changes in `VmServiceWrapper.java` are conditions encountered with the Dart VM when debugging Flutter apps, which should harden our debugger implementation, I think there are some other changes I need to migrate

**Comment by alexander-doroshko** [comment]
Thanks!
Two thoughts.

Regarding run configuration UI. As the whole process is based on the `webdev` tool, I think we should have options according to the [webdev](https://pub.dev/packages/webdev#webdev-serve) command line. As a start, directory name and port would probably be enough (that's what we effectively have in the current workflow). I think this would be better than HTML file path (which we, in fact, do not need).

We need to handle run/debug session termination better. Currently, when the user closes the Chrome window, Run/Debug session in the IDE remains visually active (and daemon remains alive). Do we want the daemon to stay alive between launches? It could make nest start faster though I'm not sure how much faster. For us, it's much easier to start/stop daemon as user starts/ends run/debug session, so we may start with this. But then we need to make sure that daemon stops on Chrome close.

Also, it would be great to see the webdev command line and 'Process finished with exit code x' in the console.

File: Dart/src/com/jetbrains/lang/dart/DartBundle.properties (hunk 2 above)
Comment (devoncarew) [reply Dart/src/com/jetbrains/lang/dart/DartBundle.properties]: Or just, 'Debug Dart web application'?

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/VmServiceWrapper.java [hunk 4]
```diff
@@ -116,6 +117,11 @@ public void received(final Isolate isolate) {
                         return;
                       }
 
+                      // This is the entry point for attaching a debugger to a running app.
+                      if (eventKind == EventKind.Resume) {
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/VmServiceWrapper.java]: Do you know how to reach this branch?

**Comment by devoncarew** [comment]
> We need to handle run/debug session termination better. Currently, when the user closes the Chrome window, Run/Debug session in the IDE remains visually active (and daemon remains alive). Do we want the daemon to stay alive between launches? It could make nest start faster though I'm not sure how much faster.

My guess is that having one webdev instance per launch makes sense.

> For us, it's much easier to start/stop daemon as user starts/ends run/debug session, so we may start with this. But then we need to make sure that daemon stops on Chrome close.

webdev should be doing this - if it isn't, we should just file a bug on https://github.com/dart-lang/webdev.

File: Dart/src/com/jetbrains/lang/dart/DartBundle.properties (hunk 2 above)
Comment (alexander-doroshko) [reply Dart/src/com/jetbrains/lang/dart/DartBundle.properties]: >Or just, 'Debug Dart web application'?

👍 That was my thought too.
Or 'Debug Dart web application using the webdev tool'

**Comment by alexander-doroshko** [comment]
> webdev should be doing this - if it isn't, we should just file a bug on https://github.com/dart-lang/webdev.

webdev server doesn't terminate on Chrome close

**Comment by jwren** [comment]
@alexander-doroshko I just uploaded a few more commits on this PR, and just saw your comments, thanks for taking a look.  I'll follow up with the Webdev devs on the Dart team to see if the chrome tab closing issue is a known issue.  The Windows issues are currently a work in progress for them.

**Comment by devoncarew** [comment]
> webdev server doesn't terminate on Chrome close (at least in my environment)

From talking to the webdev team, `webdev serve` won't terminate when the chrome process closes, but `webdev daemon` should.

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/VmServiceWrapper.java (hunk 4 above)
Comment (jwren) [reply Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/VmServiceWrapper.java]: I reverted the changes in this file.  When I saw the difference to what the Dart Plugin (as compared to the Flutter plugin) I originally thought that the changes were for webdev.  @stevemessick advice as necessary.

**Comment by jwren** [comment]
> Thanks!
> Two thoughts.
> 
> Regarding run configuration UI. As the whole process is based on the `webdev` tool, I think we should have options according to the [webdev](https://pub.dev/packages/webdev#webdev-serve) command line. As a start, directory name and port would probably be enough (that's what we effectively have in the current workflow). I think this would be better than HTML file path (which we, in fact, do not need).
> 

Having a `cwd` (current working directory) would be more in line with what the `webdev` command line would take, however it was a user experience conversation that lead me to include the `*.html` file instead.  When users go to launch some web application they think more in terms of some html file, not a directory.

As for including a port, the thought was that we could add it if users ask, but if possible, lets not include preferences were we don't need them.

@jakemac53 & @grouma -- would it be possible to have the `*.html` file be passed in as a parameter on the command line.  It seems reasonable that a user might have an `index.html`, but also some other `html` files that they might want to launch and debug.


> We need to handle run/debug session termination better. Currently, when the user closes the Chrome window, Run/Debug session in the IDE remains visually active (and daemon remains alive). Do we want the daemon to stay alive between launches? It could make the next start faster though I'm not sure how much faster. For us, it's much easier to start/stop daemon as user starts/ends run/debug session, so we may start with this. But then we need to make sure that daemon stops on Chrome close.
> 
> Also, it would be great to see the webdev command line and 'Process finished with exit code x' in the console.

@grouma  -- for context, this *is* run with `daemon` mode

**Comment by alexander-doroshko** [comment]
Thanks! I'll merge it on Monday.

File: Dart/src/com/jetbrains/lang/dart/DartBundle.properties [hunk 5]
```diff
@@ -98,10 +98,18 @@ fields.to.generate.setters=Select Fields to Generate Setters
...
 dart.implement.method=Choose method to implement
+
 runner.command.line.configuration.name=Dart Command Line App
 runner.command.line.configuration.description=Dart Command Line Application
+
 remote.debug.configuration.name=Dart Remote Debug
 remote.debug.configuration.description=Debug Remote Dart Command Line Application
+
+webdev.debug.configuration.name=Dart Web
+webdev.debug.configuration.description=Debug Dart Webdev
```
Comment (devoncarew) [inline Dart/src/com/jetbrains/lang/dart/DartBundle.properties]: `Debug Dart Web App`?

File: Dart/src/com/jetbrains/lang/dart/DartBundle.properties [hunk 6]
```diff
@@ -98,10 +98,18 @@ fields.to.generate.setters=Select Fields to Generate Setters
...
 dart.override.method=Choose method to override
 dart.implement.method=Choose method to implement
+
 runner.command.line.configuration.name=Dart Command Line App
 runner.command.line.configuration.description=Dart Command Line Application
+
 remote.debug.configuration.name=Dart Remote Debug
 remote.debug.configuration.description=Debug Remote Dart Command Line Application
+
+webdev.debug.configuration.name=Dart Web
```
[the remark this answers is not in the document]
Comment (jwren) [reply Dart/src/com/jetbrains/lang/dart/DartBundle.properties]: It originally included 'App' (before my first cut of this PR), I am happy to include or exclude it.  Flutter has 'Flutter' not 'Flutter App'..., it felt redundant, but again, I can be convinced either way.

File: Dart/src/com/jetbrains/lang/dart/DartBundle.properties (hunk 5 above)
Comment (jwren) [reply Dart/src/com/jetbrains/lang/dart/DartBundle.properties]: Same thought here as above.

File: Dart/src/com/jetbrains/lang/dart/DartBundle.properties [hunk 7]
```diff
@@ -98,10 +98,18 @@ fields.to.generate.setters=Select Fields to Generate Setters
...
+webdev.debug.configuration.description=Debug Dart Webdev
+choose.html.main.file=Choose HTML File
+old.dart.sdk.for.webdev=Dart SDK {0}+ is required for Dart webdev debugging, current version: {1}
```
[the remark this answers is not in the document]
Comment (jwren) [reply Dart/src/com/jetbrains/lang/dart/DartBundle.properties]: This message appears in the run config parameter error/ exception section which has limited space.  I'll iterate.

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import org.jetbrains.annotations.NotNull;
+
+public class DartWebdevConfigurationType extends ConfigurationTypeBase implements DumbAware {
+  @NotNull
+  public static DartWebdevConfigurationType getInstance() {
+    return ConfigurationTypeUtil.findConfigurationType(DartWebdevConfigurationType.class);
+  }
+
+  public DartWebdevConfigurationType() {
+    super("DartRemoteDebugConfigurationType",
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConfigurationType.java]: copy/paste typo :)

File: Dart/src/com/jetbrains/lang/dart/DartBundle.properties (hunk 6 above)
[the remark this answers is not in the document]
Comment (devoncarew) [reply Dart/src/com/jetbrains/lang/dart/DartBundle.properties]: Gotcha. Looking at other run configurations, it looks like they mostly omit any trailing 'App' text.

**Comment by alexander-doroshko** [comment]
Sorry, I failed to merge it today but will definitely merge tomorrow and we'll iterate further.

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConsoleView.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    super(project, searchScope, true, false);
+  }
+
+  @Override
+  public void print(@NotNull String text, @NotNull ConsoleViewContentType contentType) {
+    if (!text.startsWith("[{")) {
+      super.print(text, contentType);
+    }
+    final String logMessage = DartDaemonParserUtil.getLogMessage(text.trim());
+    if (logMessage != null && !logMessage.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (jwren) [reply Dart/src/com/jetbrains/lang/dart/ide/runner/server/webdev/DartWebdevConsoleView.java]: It is part of the log coming out of `webdev'.

CC @grouma

**Comment by jwren** [comment]
Thanks for the comments @devoncarew.

**Comment by alexander-doroshko** [comment]
Merged with some minor tweaks.

**Pull request closed without merging**